go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,317 @@
+			channel, err := mcap.ParseChannel(data)
+			if err != nil {
+				return err
+			}
+			if len(opts.includeTopics) > 0 {
+				for _, matcher := range opts.includeTopics {
+					if matcher.MatchString(channel.Topic) {
+						channels[channel.ID] = markableChannel{channel, false}
+					}
+				}
+			} else if len(opts.excludeTopics) > 0 {
+				for _, matcher := range opts.excludeTopics {
+					if !matcher.MatchString(channel.Topic) {
+						channels[channel.ID] = markableChannel{channel, false}
+					}
+				}
+			} else {
+				channels[channel.ID] = markableChannel{channel, false}
+			}
+		case mcap.TokenMessage:
+			message, err := mcap.ParseMessage(data)
+			if err != nil {
+				return err
+			}
+			if message.LogTime < opts.start {
+				continue
+			}
+			if message.LogTime >= opts.end {
+				continue
+			}
+			channel, ok := channels[message.ChannelID]
+			if !ok {
+				continue
+			}
+			if !channel.written {
+				schema, ok := schemas[channel.SchemaID]
+				if !ok {
+					return fmt.Errorf("encountered channel with topic %s with unknown schema ID %d", channel.Topic, channel.SchemaID)
+				}
+				if !schema.written {
+					if err = mcapWriter.WriteSchema(schema.Schema); err != nil {
+						return err
+					}
+					schema.written = true
+				}
+				if err = mcapWriter.WriteChannel(channel.Channel); err != nil {
+					return err
+				}
+				channel.written = true
+			}
+			if err = mcapWriter.WriteMessage(message); err != nil {
+				return err
+			}
+		case mcap.TokenAttachment:
+			if !opts.includeAttachments {
+				continue
+			}
+			attachment, err := mcap.ParseAttachment(data)
+			if err != nil {
+				return err
+			}
+			if attachment.LogTime < opts.start {
+				continue
+			}
+			if attachment.LogTime >= opts.end {
+				continue
+			}
+			if err = mcapWriter.WriteAttachment(attachment); err != nil {
+				return err
+			}
+		case mcap.TokenMetadata:
+			if !opts.includeMetadata {
+				continue
+			}
+			metadata, err := mcap.ParseMetadata(data)
+			if err != nil {
+				return err
+			}
+			if err = mcapWriter.WriteMetadata(metadata); err != nil {
+				return err
+			}
+		case mcap.TokenDataEnd, mcap.TokenFooter:
+			// data section is over, either because the file is over or the summary section starts.
+			return nil
+		case mcap.TokenChunk:
+			return errors.New("expected lexer to remove chunk records from input stream")
+		case mcap.TokenError:
+			return errors.New("received error token but lexer did not return error on Next")
+		}
+	}
+}
+
+func init() {
+	rootCmd.AddCommand(filterCmd)
+
+	filterCmd.PersistentFlags().StringVarP(&filterOutput, "output", "o", "", "output filename")
+	filterCmd.PersistentFlags().StringArrayVarP(&filterIncludeTopics, "include-topic-regex", "y", []string{}, "messages with topic names matching this regex will be included")
+	filterCmd.PersistentFlags().StringArrayVarP(&filterExcludeTopics, "exclude-topic-regex", "n", []string{}, "messages with topic names matching this regex will be included")
```

Comment:
maybe we can indicate somehow that multiple of these can be supplied? might need to futz with it to keep the wording short